src/server/pkg/hashtree/db.go

```diff
@@ -0,0 +1,1703 @@
+			Offset: w.offset,
+		})
+	}
+	b, err := w.pbw.WriteBytes(n.k)
+	if err != nil {
+		return err
+	}
+	w.offset += uint64(b)
+	b, err = w.pbw.WriteBytes(n.v)
+	if err != nil {
+		return err
+	}
+	w.offset += uint64(b)
+	return nil
+}
+
+// Copy copies a hashtree reader in a writer.
+func (w *Writer) Copy(r *Reader) error {
+	for {
+		n, err := r.Read()
+		if err != nil {
+			if err == io.EOF {
+				return nil
+			}
+			return err
+		}
+		if err := w.Write(n); err != nil {
+			return err
+		}
+	}
+}
+
+// Index returns the index for a hashtree writer.
+func (w *Writer) Index() ([]byte, error) {
+	buf := &bytes.Buffer{}
+	pbw := pbutil.NewWriter(buf)
+	for _, idx := range w.idxs {
+		if _, err := pbw.Write(idx); err != nil {
+			return nil, err
+		}
+	}
+	return buf.Bytes(), nil
+}
+
+// GetRangeFromIndex returns a subtree byte range in a serialized hashtree based on a passed in prefix.
+func GetRangeFromIndex(r io.Reader, prefix string) (uint64, uint64, error) {
+	prefix = clean(prefix)
+	pbr := pbutil.NewReader(r)
+	idx := &Index{}
+	k := b(prefix)
+	var lower, upper uint64
+	iter := func(f func(int) bool) error {
+		for {
+			if err := pbr.Read(idx); err != nil {
+				if err == io.EOF {
+					break
+				}
+				return err
+			}
+			var cmp int
+			if len(k) < len(idx.K) {
+				cmp = bytes.Compare(k, idx.K[:len(k)])
+			} else {
+				cmp = bytes.Compare(k[:len(idx.K)], idx.K)
+			}
+			if f(cmp) {
+				break
+			}
+		}
+		return nil
+	}
+	low := func(cmp int) bool {
+		if cmp > 0 {
+			lower = idx.Offset
+			return false
+		} else if cmp < 0 {
+			// Handles the case where a prefix fits within one range
+			upper = idx.Offset
+		}
+		return true
+	}
+	up := func(cmp int) bool {
+		if cmp < 0 {
+			upper = idx.Offset
+			return true
+		}
+		return false
+	}
+	// Find lower
+	iter(low)
+	// Find upper
+	if upper <= 0 {
+		iter(up)
+	}
+	// Handles the case when at the end of the indexes
+	if upper <= 0 {
+		return lower, 0, nil
+	}
+	// Return offset and size
+	return lower, upper - lower, nil
+}
+
+// NewFilter creates a filter for a hashtree shard.
+func NewFilter(numTrees int64, tree int64) func(k []byte) (bool, error) {
+	return func(k []byte) (bool, error) {
+		if pathToTree(k, numTrees) == uint64(tree) {
+			return true, nil
+		}
+		return false, nil
+	}
+}
+
+// PathToTree computes the hashtree shard for a path.
+func PathToTree(path string, numTrees int64) uint64 {
+	path = clean(path)
+	return pathToTree(b(path), numTrees)
+}
+
+func pathToTree(k []byte, numTrees int64) uint64 {
+	return xxhash.Checksum64(k) % uint64(numTrees)
+}
+
+type nodeStream struct {
+	node *MergeNode
+	r    *Reader
+}
+
+type mergePQ struct {
+	q    []*nodeStream
+	size int
+}
+
+func (mq *mergePQ) k(i int) []byte {
+	return mq.q[i].node.k
+}
+
+func (mq *mergePQ) insert(s *nodeStream) error {
+	// Get next node in stream
+	var err error
+	s.node, err = s.r.Read()
+	if err != nil {
+		if err == io.EOF {
+			return nil
+		}
+		return err
+	}
+	mq.q[mq.size+1] = s
```

Comment:
How do we know that `mq.size+1` isn't out of bounds here? I don't see where `mq.q` gets initialized. Could it be an `append` instead?

Reply:
https://github.com/pachyderm/pachyderm/blob/f07bb15aed7cd137d0f5920b3c551f24ec66a45f/src/server/pkg/hashtree/db.go#L1195

So, I do not think append would work here because getting a node from the queue does not change the queue's slice size. The queue size change is reflected in the size field. I could make a change to the fill function to do that, but I think this is the more performant solution because I know how big the slice can get from the beginning.